runtime/jcl/common/jclvm.c

```diff
@@ -138,6 +150,146 @@ Java_com_ibm_oti_vm_VM_allInstances(JNIEnv * env, jclass unused, jclass clazz, j
 	return count;
 }
 
+/**
+ * Return a String object containing a summary of the classes on the heap, 
+ * the number of instances, and their aggregate size.
+ */
+jobject JNICALL
+Java_openj9_tools_attach_diagnostics_base_DiagnosticUtils_getHeapClassStatisticsImpl(JNIEnv * env, jclass unused)
+{
+	J9VMThread *vmThread = (J9VMThread *) env;
+	J9JavaVM *vm = vmThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9HashTable *statsTable = NULL;
+	J9HashTableState hashTableState;
+	BOOLEAN outOfMemory = FALSE;
+	j9object_t stringObject = NULL;
+	J9HeapStatisticsTableEntry **statsArray = NULL;
+	jobject stringObjectRef = NULL;
+	I_32 numClasses = 0;
+
+	PORT_ACCESS_FROM_ENV(env);
+
+	vmFuncs->internalEnterVMFromJNI(vmThread);
+
+	vmFuncs->acquireExclusiveVMAccess(vmThread);
+	statsTable = collectHeapStatistics(vmThread);
+	vmFuncs->releaseExclusiveVMAccess(vmThread);
+
+	if (NULL != statsTable) {
+		numClasses = hashTableGetCount(statsTable);
+		statsArray = j9mem_allocate_memory(numClasses * sizeof(J9HeapStatisticsTableEntry*), J9MEM_CATEGORY_VM_JCL);
+	}
+	if (NULL == statsArray) {
+		outOfMemory = TRUE;
+	} else {
+		I_32 cursor = 0;
+		J9HeapStatisticsTableEntry *entry = (J9HeapStatisticsTableEntry *) hashTableStartDo(statsTable, &hashTableState);
+		while (NULL != entry) {
+			entry->aggregateSize = entry->objectSize * entry->objectCount;
+			statsArray[cursor] = entry;
+			cursor += 1;
+			entry = (J9HeapStatisticsTableEntry *) hashTableNextDo(&hashTableState);
+		}
+		hashTableFree(statsTable);
+		numClasses = cursor; /* adjust the length in case the hash table contained nulls */
+		sortStatsTable(statsArray, 0, numClasses - 1);
+	}
+	if (!outOfMemory) {
+		I_32 printedLength = 0;
+		UDATA bufferSize =0;
+		do {
+			bufferSize += numClasses * 80; /* try incrementally larger sizes */
+			char *stringBuffer = (char *) j9mem_allocate_memory(bufferSize, J9MEM_CATEGORY_VM_JCL);
+			if (NULL == stringBuffer) {
+				outOfMemory = TRUE;
+				break;
+			} else {
```

Comment:
Removing the `else` and its block would reduce nesting.

Reply:
Good idea.